### PR TITLE
pkg/ccn-lite/relic: check for minimal cmake version

### LINKED ISSUE #9064: examples: unable to build ccn-lite for cortex-m
#### Description
Just tried to build the `cc-lite-relay` example for ARM Cortex-M (`iotlab-m3` in my case). The build fails due to some libc issues as it seems to me:

```
hauke@Prag ~/dev/riot/RIOT/examples/ccn-lite-relay $ BOARD=iotlab-m3 make all
Building application "ccn-lite-relay" for "iotlab-m3" with MCU "stm32f1".

make[1]: Nothing to be done for 'prepare'.
make[1]: Nothing to be done for 'prepare'.
"make" -C /home/hauke/dev/riot/RIOT/pkg/ccn-lite
/home/hauke/dev/riot/RIOT/dist/tools/cmake/generate-xcompile-toolchain.sh > /home/hauke/dev/riot/RIOT/examples/ccn-lite-relay/bin/pkg/iotlab-m3/ccn-lite/xcompile-toolchain.cmake
cd /home/hauke/dev/riot/RIOT/examples/ccn-lite-relay/bin/pkg/iotlab-m3/ccn-lite/src && \
cmake -DCMAKE_TOOLCHAIN_FILE=/home/hauke/dev/riot/RIOT/examples/ccn-lite-relay/bin/pkg/iotlab-m3/ccn-lite/xcompile-toolchain.cmake \
	  -DCCNL_RIOT=1 -DRIOT_CFLAGS="-Wall -Werror -Wextra -Wno-implicit-fallthrough -mno-thumb-interwork -mcpu=cortex-m3 -mlittle-endian -mthumb -mfloat-abi=soft -ffunction-sections -fdata-sections -fno-builtin -fshort-enums -ggdb -g3 -Os -std=c99 -fno-delete-null-pointer-checks -fdiagnostics-color -Wstrict-prototypes -Wold-style-definition -Werror=strict-prototypes -Werror=old-style-definition -fno-common -Wall -include '/home/hauke/dev/riot/RIOT/examples/ccn-lite-relay/bin/iotlab-m3/riotbuild/riotbuild.h'  -isystem /newlib-nano -I/home/hauke/dev/riot/RIOT/core/include -I/home/hauke/dev/riot/RIOT/drivers/include -I/home/hauke/dev/riot/RIOT/sys/include -I/home/hauke/dev/riot/RIOT/cpu/stm32f1/include -I/home/hauke/dev/riot/RIOT/boards/iotlab-m3/include -I/home/hauke/dev/riot/RIOT/boards/common/iotlab/include -I/home/hauke/dev/riot/RIOT/boards/iotlab-m3/include -I/home/hauke/dev/riot/RIOT/cpu/stm32_common/include -I/home/hauke/dev/riot/RIOT/cpu/cortexm_common/include -I/home/hauke/dev/riot/RIOT/cpu/cortexm_common/include/vendor -I/home/hauke/dev/riot/RIOT/sys/libc/include -I/home/hauke/dev/riot/RIOT/examples/ccn-lite-relay/bin/pkg/iotlab-m3/ccn-lite/src/ccnl-riot/include -I/home/hauke/dev/riot/RIOT/examples/ccn-lite-relay/bin/pkg/iotlab-m3/ccn-lite/src/ccnl-core/include -I/home/hauke/dev/riot/RIOT/examples/ccn-lite-relay/bin/pkg/iotlab-m3/ccn-lite/src/ccnl-pkt/include -I/home/hauke/dev/riot/RIOT/examples/ccn-lite-relay/bin/pkg/iotlab-m3/ccn-lite/src/ccnl-fwd/include -I/home/hauke/dev/riot/RIOT/sys/posix/include -I/home/hauke/dev/riot/RIOT/examples/ccn-lite-relay/bin/pkg/iotlab-m3/tlsf/src -I/home/hauke/dev/riot/RIOT/drivers/at86rf2xx/include" -DBUILD_TESTING=OFF .
-- The CXX compiler identification is GNU 7.2.1
-- Check for working CXX compiler: /home/hauke/tools/bin/arm-none-eabi-g++
-- Check for working CXX compiler: /home/hauke/tools/bin/arm-none-eabi-g++ -- broken
CMake Error at /usr/share/cmake-3.5/Modules/CMakeTestCXXCompiler.cmake:54 (message):
  The C++ compiler "/home/hauke/tools/bin/arm-none-eabi-g++" is not able to
  compile a simple test program.

  It fails with the following output:

   Change Dir: /home/hauke/dev/riot/RIOT/examples/ccn-lite-relay/bin/pkg/iotlab-m3/ccn-lite/src/CMakeFiles/CMakeTmp

  

  Run Build Command:"/usr/bin/make" "cmTC_22036/fast"

  make[2]: Entering directory
  '/home/hauke/dev/riot/RIOT/examples/ccn-lite-relay/bin/pkg/iotlab-m3/ccn-lite/src/CMakeFiles/CMakeTmp'


  /usr/bin/make -f CMakeFiles/cmTC_22036.dir/build.make
  CMakeFiles/cmTC_22036.dir/build

  make[3]: Entering directory
  '/home/hauke/dev/riot/RIOT/examples/ccn-lite-relay/bin/pkg/iotlab-m3/ccn-lite/src/CMakeFiles/CMakeTmp'


  Building CXX object CMakeFiles/cmTC_22036.dir/testCXXCompiler.cxx.obj

  /home/hauke/tools/bin/arm-none-eabi-g++ -o
  CMakeFiles/cmTC_22036.dir/testCXXCompiler.cxx.obj -c
  /home/hauke/dev/riot/RIOT/examples/ccn-lite-relay/bin/pkg/iotlab-m3/ccn-lite/src/CMakeFiles/CMakeTmp/testCXXCompiler.cxx


  Linking CXX executable cmTC_22036

  /usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_22036.dir/link.txt
  --verbose=1

  /home/hauke/tools/bin/arm-none-eabi-g++
  CMakeFiles/cmTC_22036.dir/testCXXCompiler.cxx.obj -o cmTC_22036

  
  /home/hauke/tools/gcc-arm-none-eabi-7-2017-q4-major/bin/../lib/gcc/arm-none-eabi/7.2.1/../../../../arm-none-eabi/lib/libc.a(lib_a-exit.o):
  In function `exit':

  exit.c:(.text.exit+0x2c): undefined reference to `_exit'

  collect2: error: ld returned 1 exit status

  CMakeFiles/cmTC_22036.dir/build.make:97: recipe for target 'cmTC_22036'
  failed

  make[3]: *** [cmTC_22036] Error 1

  make[3]: Leaving directory
  '/home/hauke/dev/riot/RIOT/examples/ccn-lite-relay/bin/pkg/iotlab-m3/ccn-lite/src/CMakeFiles/CMakeTmp'


  Makefile:126: recipe for target 'cmTC_22036/fast' failed

  make[2]: *** [cmTC_22036/fast] Error 2

  make[2]: Leaving directory
  '/home/hauke/dev/riot/RIOT/examples/ccn-lite-relay/bin/pkg/iotlab-m3/ccn-lite/src/CMakeFiles/CMakeTmp'


  

  

  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  ccnl-core/CMakeLists.txt:2 (project)


-- Configuring incomplete, errors occurred!
See also "/home/hauke/dev/riot/RIOT/examples/ccn-lite-relay/bin/pkg/iotlab-m3/ccn-lite/src/CMakeFiles/CMakeOutput.log".
See also "/home/hauke/dev/riot/RIOT/examples/ccn-lite-relay/bin/pkg/iotlab-m3/ccn-lite/src/CMakeFiles/CMakeError.log".
Makefile:17: recipe for target '/src/Makefile' failed
make[1]: *** [/src/Makefile] Error 1
/home/hauke/dev/riot/RIOT/examples/ccn-lite-relay/../../Makefile.include:418: recipe for target '/home/hauke/dev/riot/RIOT/examples/ccn-lite-relay/bin/iotlab-m3/ccn-lite.a' failed
make: *** [/home/hauke/dev/riot/RIOT/examples/ccn-lite-relay/bin/iotlab-m3/ccn-lite.a] Error 2
```

using 
```
arm-none-eabi-g++ (GNU Tools for Arm Embedded Processors 7-2017-q4-major) 7.2.1 20170904 (release) [ARM/embedded-7-branch revision 255204]
```

Is this something known (and I was to blind to find the issue), or did someone encounter this before?

#### Steps to reproduce the issue
build the `ccn-lite-relay` example for `iotlab-m3` using the toolchain as pointed to above.

#### Expected results
should build without errors...

#### Actual results
see above

#### Versions
see above

### PR DESCRIPTION
**Pull request by cladmi**
### Contribution description

ccn-lite/relic do not build with cmake < 3.6.0. This checks for a minimal version
instead of failing to compile.

    cmake version 3.5.2 is not >= to minimal required 3.6.0

    Makefile:25: recipe for target '..cmake_version_supported' failed

Tested with versions 3.5.2, 3.6.0-rc1, 3.6.0, 3.6.3, 3.7.0-rc1, 3.7.0.

Note: the check used does not consider '-rcX' as 'sort -V' does not handle them
properly.

### Testing

You can test this by compiling `examples/ccn-lite-relay` with different cmake versions:

Last 3.5 release

```
PATH=/tmp/cmake/cmake-3.5.2-Linux-x86_64/bin/:${PATH} make clean all BOARD=samr21-xpro

cmake version 3.5.2 is not >= to minimal required 3.6.0

Makefile:28: recipe for target '..cmake_version_supported' failed
make[1]: *** [..cmake_version_supported] Error 1
```

First 3.6 stable release

```
PATH=/tmp/cmake/cmake-3.6.0-Linux-x86_64/bin/:${PATH} make clean all BOARD=samr21-xpro
...
Build successfully
```

You can download cmake old versions here:

https://cmake.org/files/

### Issues/PRs references

Fixes #9064 by making the problem clear

Found during release testing.

### CODE REVIEW

**Comment by cladmi** [comment]
@smlng I asked you to also have a test run on `mac`.

**Comment by miri64** [comment]
@cladmi, some packet used in `unittests` causes a similar issue. Can you check?

**Comment by cladmi** [comment]
I thought about it just after submitting it ><

File: pkg/ccn-lite/Makefile [hunk 1]
```diff
@@ -13,11 +16,22 @@ all: $(PKG_BUILDDIR)/src/Makefile
 	$(MAKE) -C $(PKG_BUILDDIR)/src && \
 	cp $(PKG_BUILDDIR)/src/lib/libccnl-riot.a $(BINDIR)/ccn-lite.a
 
-$(PKG_BUILDDIR)/src/Makefile: $(TOOLCHAIN_FILE)
+$(PKG_BUILDDIR)/src/Makefile: $(TOOLCHAIN_FILE) | ..cmake_version_supported
```
Comment (cgundogan) [inline pkg/ccn-lite/Makefile]: This is executed with every run without being cached, right? Can't we touch a file to skip subsequent version checks for an existing binary dir?

Comment (cladmi) [reply pkg/ccn-lite/Makefile]: Here `TOOLCHAIN_FILE` is also rebuilt everytime as it depends on `git-download` and so is `Makefile` created by `cmake`…

A solution would still require to save `command -v cmake` to a file and verify it did not change since last call.
So I would more do it after I can finish working on https://github.com/RIOT-OS/RIOT/pull/9634.

Comment (smlng) [reply pkg/ccn-lite/Makefile]: can't you run the cmake version check *before* downloading or generating the toolchain file?

**Comment by cladmi** [comment]
I have a quoting issue in the command, does this work ?

    cmake --version | sed -n '1 {s/cmake version //;s/-rc.*//;p}'

File: pkg/ccn-lite/Makefile (hunk 1 above)
Comment (cladmi) [reply pkg/ccn-lite/Makefile]: I was more concentrated on getting a working test than where I should put it.

Putting it as a check before downloading could be problematic if `download` gets done at a separate time or even a different environment. I still hope for a separate "prepare" step. But it could be moved again later.

So yes I can move it.

**Comment by smlng** [comment]
@cladmi the `p` command must be closed with a `;` too. Because this works for me:

```
cmake --version | sed -n '1 {s/cmake version //;s/-rc.*//;p;}'
```

**Comment by cladmi** [comment]
The issue for `unittests` is in `relic` that has the same issue.

I will move the test to a dedicated script to allow reusing it.

**Comment by cladmi** [comment]
Problem when doing the check before downloading, is that `pkg-prepare` errors are ignored:

```
pkg-prepare:
	-@for i in $(USEPKG) ; do "$(MAKE)" -C $(RIOTPKG)/$$i prepare ; done
```

So the error message is shown two times…

**Comment by cladmi** [comment]
It still prevents to download the source though so could be kept this way.

File: dist/tools/has_higher_version/has_higher_version.sh [hunk 2]
```diff
@@ -0,0 +1,33 @@
+#! /bin/bash
+#
+# usage: has_higher_version.sh <version> <minimal_version> [toolname]
```
Comment (smlng) [inline dist/tools/has_higher_version/has_higher_version.sh]: wouldn't `has_minimal_version` be semantically (more) correct because the version doesn't have to be higher/greater but greater *or* equal, hence it should at least match the *minimal* version given

File: dist/tools/has_higher_version/has_higher_version.sh [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  readonly TOOLSTR="${TOOLNAME} "
+else
+  readonly TOOLSTR=""
+fi
+
+
+HIGHEST=$(printf "%s\n%s\n" "${TOOLVERSION}" "${MINVERSION}" | sort -rV | head -n 1)
+test "${HIGHEST}" = "${TOOLVERSION}" && exit 0
+
+printf "\n%sversion %s is not >= to minimal required %s\n\n" "${TOOLSTR}" "${TOOLVERSION}" "${MINVERSION}"
```
Comment (smlng) [inline dist/tools/has_higher_version/has_higher_version.sh]: maybe rephrase: `printf "\nInvalid version, found %s%s, minimal required is %s\n\n" "${TOOLSTR}" "${TOOLVERSION}" "${MINVERSION}"`

The `>=` should be implied by *minimal required*.

File: dist/tools/has_higher_version/has_higher_version.sh [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+if [ "${TOOLNAME}" != "" ]; then
+  readonly TOOLSTR="${TOOLNAME} "
+else
+  readonly TOOLSTR=""
+fi
+
+
+HIGHEST=$(printf "%s\n%s\n" "${TOOLVERSION}" "${MINVERSION}" | sort -rV | head -n 1)
+test "${HIGHEST}" = "${TOOLVERSION}" && exit 0
```
[the remark this answers is not in the document]
Comment (cladmi) [reply dist/tools/has_higher_version/has_higher_version.sh]: I would not output it in the build system. We say nothing about "gcc has a good version", "make is up to date", "wget is compatible".
Or I could print something to standard output and silent the error in Make.

File: dist/tools/has_higher_version/has_higher_version.sh (hunk 3 above)
Comment (cladmi) [reply dist/tools/has_higher_version/has_higher_version.sh]: Will do, I should also print to stderr.

File: dist/tools/has_higher_version/has_higher_version.sh (hunk 2 above)
Comment (cladmi) [reply dist/tools/has_higher_version/has_higher_version.sh]: I will rename the scripts.

Comment (cladmi) [reply dist/tools/has_higher_version/has_higher_version.sh]: I also used "minimal version" everywhere but was not able to name the script accordingly ><

**Comment by cladmi** [comment]
Updated. Tell me if you want an output that I silent in make.

**Comment by cladmi** [comment]
Can I squash ?

**Comment by smlng** [comment]
yes please!

**Comment by cladmi** [comment]
The rebase has the same diff as before, except that ccn-lite version was bumped in between.

I retested compiling `ccn-lite-relay` on IoT-LAB with `cmake version 3.0.2` and on my computer with `cmake version 3.11.4`.
